packages/core/crates/core-misc/src/environment.rs

```diff
@@ -0,0 +1,265 @@
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct ProtocolConfig {
+    environments: std::collections::HashMap<String, Environment>,
+    networks: std::collections::HashMap<String, NetworkOptions>,
+}
+
+/// Reads current protocol config and returns it
+pub fn get_protocol_config(mono_repo_path: &str) -> Result<ProtocolConfig, JsValue> {
+    let protocol_config_path = format!("{}/packages/core/protocol-config.json", mono_repo_path);
+
+    let data = real::read_file(protocol_config_path.as_str())?;
+
+    serde_json::from_slice::<ProtocolConfig>(&data).map_err(|e| JsValue::from(e.to_string()))
+}
+
+#[derive(Deserialize)]
+struct PackageJsonFile {
+    version: String,
+}
+
+/// Reads `hoprd` package.json file and returns its version
+fn get_package_version(mono_repo_path: &str) -> Result<String, JsValue> {
+    let package_json_path = format!("{}/packages/hoprd/package.json", mono_repo_path);
+
+    let data = real::read_file(package_json_path.as_str())?;
+
+    let package_json = serde_json::from_slice::<PackageJsonFile>(&data)
+        .map_err(|e| JsValue::from(e.to_string()))?;
+
+    /*
+     * Coerced full version using
+     * coerce_version('42.6.7.9.3-alpha') // '42.6.7'
+     */
+    coerce_version(package_json.version)
+}
+
+/// Returns a list of environments which the node is able to work with
+pub fn supported_environments(mono_repo_path: &str) -> Result<Vec<Environment>, JsValue> {
+    let mut protocol_config_path = get_protocol_config(&mono_repo_path)?;
+    let version = get_package_version(mono_repo_path)?;
+
+    let mut allowed: Vec<Environment> = vec![];
+
+    for (id, env) in protocol_config_path.environments.iter_mut() {
+        let range = env.version_range.to_owned();
+
+        if satisfies(version.to_owned(), range).unwrap() {
+            env.id = id.to_owned();
+            allowed.push(env.to_owned())
+        }
+    }
+
+    Ok(allowed)
+}
+/// Returns the environment details, returns an error if environment is not supported
+pub fn resolve_environment(
+    mono_repo_path: &str,
+    environment_id: &str,
+    maybe_custom_provider: Option<&str>,
+) -> Result<ResolvedEnvironment, JsValue> {
+    let mut protocol_config = get_protocol_config(mono_repo_path)?;
+    let version = get_package_version(mono_repo_path)?;
+
+    let environment = protocol_config
+        .environments
+        .get_mut(environment_id)
+        .ok_or(JsValue::from(format!(
+            "Could not find environment {} in protocol config",
+            environment_id
+        )))?;
+
+    let network = protocol_config
+        .networks
+        .get_mut(&environment.network_id)
+        .ok_or(JsValue::from(format!(
+            "Invalid network_id {} for environment {}",
+            environment.network_id, environment_id
+        )))?;
+
+    network.id = environment.network_id.to_owned();
+
+    if let Some(custom_provider) = maybe_custom_provider {
+        network.default_provider = String::from(custom_provider);
```

Comment:
Also here, if LHS type is known at compile time, `custom_provider.into()` should be preferred.